src/torchio/data/queue.py

```diff
@@ -136,13 +142,53 @@ class Queue(Dataset):
     ...         targets = patches_batch['brain'][tio.DATA]  # key 'brain' is in subject
     ...         logits = model(inputs)  # model being an instance of torch.nn.Module
 
+
+    Example:
+
+    >>> # Usage with distributed training
+    >>> import torch.distributed as dist
+    >>> from torch.utils.data.distributed import DistributedSampler
+    >>> # Assume a process running on distributed node 3
+    >>> rank = 3
+    >>> patch_sampler = tio.data.UniformSampler(patch_size)
+    >>> subject = tio.datasets.Colin27()
+    >>> subjects_dataset = tio.SubjectsDataset(10 * [subject])
+    >>> subject_sampler = dist.DistributedSampler(
+    ...     subjects_dataset,
+    ...     rank=local_rank,
+    ...     shuffle=True,
+    ...     drop_last=True,
```

Comment:
Would you recommend setting this to `True`?

Reply:
The shuffle keyword defaults to True for dist.DistributedSampler, so you can use it as is. I think your question is about drop_last.

When training with DDP, backward is promised to be handled in sync with each other. The processes in each rank don't know which iteration of which epoch they are processing, they just check that all other processes have entered the backward step.

Without drop_last, the number of iterations a process has to process when the number of subjects is not divided and dropped will vary and become unmanageable.

For example. You have 3 GPUs and rank 0-2 using DDP. The number of subjects is 31. With drop_last=False, rank 0 is allocated 11 subjects, rank 1 is allocated 10 subjects, and rank 2 is allocated 10 subjects. Watch carefully as rank 0 patches the 11th subject, puts it into the model, and tries to backward. Since ranks 1 and 2 do not have the 11th subject, they move on to the next epoch and start learning.

So rank 0 and rank 1 and 2 are learning a little bit out of sync, which is not what we wanted. The scary part is that the program doesn't show any errors. The programmer will be surprised by the sync time out error that suddenly appears after a very long period of learning. Worse yet, the program will finish learning completely, but just perform poorly. In this case, it's really hard to find the problem.

In conclusion, I think drop_last should be set to default True.

Translated with www.DeepL.com/Translator (free version)

Reply:
Thanks for the detailed explanation, makes sense.